### ISSUE TITLE
Code highlighting activated in .Rnw files after using \Sexpr with square brackets

### ISSUE THREAD

**Issue by nickreich** [rstudio/rstudio#14382]
### System details

    RStudio Edition : Desktop
    RStudio Version : Version 2023.12.1+402 (2023.12.1+402)
    OS Version      : 12.7.3
    R Version       : 4.3.2

### Steps to reproduce the problem

I cannot attach a .Rnw file to the issue, but I have pasted in the complete text of the file saved as .Rnw that shows the problem.

```
\documentclass{article}

\begin{document}

<<>>=
my_data <- 1:5
@

The value of my data is \Sexpr{my_data}.

for example, you can see how the word "for" at the beginning of this sentence is not color-coded.

However, now we use \Sexpr{my_data[2]}.

for example, you can see how the word "for" at the beginning of this sentence is coded as if it were code (same with "if").

And if we put a quote mark here' now everything after it is seen as code quote'.

This appears to be a bug, and if there is an ending single quote the end of document is not read when knitted directly by the IDE. Although knitting the document via code like

knitr::knit('test.Rnw')

does work.

'

\end{document}
```

### Describe the problem in detail

This problem with the RStudio IDE involves working with .Rnw or Sweave files and a specific use of `\Sexpr{}`, which is the tag for inline code. I have found that a use of `\Sexpr{}` that includes square brackets, such as `\Sexpr{my_data[2]}`, somehow activates the IDE to turn on code highlighting for all text after the square brackets. The file still does compile but the code highlighting is broken.

In the following screenshot you can see how on line 15 the word "for" is coded as text (while on line 11 it is not), and other parts of the document after apostrophes are also "commented out" in green text (including the \end{document} tag on line 25:
![image](https://github.com/rstudio/rstudio/assets/1280767/8af2605d-7959-473e-8f29-6b95a146d3be)

### Describe the behavior you expected
 The expected behavior is that code highlighting should never be turned on.

- [ x ] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [ x ] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.posit.co/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [ x ] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by kevinushey** [comment]
Thank you for the bug report! I was able to reproduce; I'll take a look.

**Comment by ronblum** [comment]
I'm unable to reproduce the the problem in earlier releases, so I can't verify this. @gtritchie Are you able to reproduce this, by chance?

**Comment by nickreich** [comment]
I think this could be closed? per #14384

**Comment by gtritchie** [comment]
> I think this could be closed? per #14384

Yes, we are working on validating the fix.

**Comment by ronblum** [comment]
Turned on rainbow parenthesis, and now I can reproduce the original problem and verify fixed in
- RStudio Desktop 2024.04.0-daily+581 on MacOS 14.4
- Posit Workbench 2024.04.0-daily+609.pro1 on Ubuntu 22.04 via Chrome on MacOS 14.4
